Simplify settings

TkTech opened the pull request:
This is here for discussion, seems like you have a way you like to do settings and I wanted to offer a counterpoint.

This uses Pydantic for better settings validation, since pydantic was already a dependency.

Pros:

- Better error messages
- .venv file support (just make a .venv in your local folder, no need to rebuild docker images)
- No repetition with prefixes
- Configure database, email, storage with DSNs

Email is configured like so:

- `TAKAHE_EMAIL_BACKEND: "console://console"`
- `TAKAHE_EMAIL_BACKEND: "smtp://apikey:<api_key>@smtp.sendgrid.net:587?tls=true"`
- `TAKAHE_EMAIL_BACKEND: "smtp://apikey:<api_key>@smtp.sendgrid.net:587?ssl=true"`

Storage is configured like so:

- `TAKAHE_STORAGE_BACKEND: "s3://<key_id>:<key secret>@s3.us-east-2.amazonaws.com/<bucket name>`
- `TAKAHE_STORAGE_BACKEND: "gcs://google/<bucket name>`

Unfortunately you have to specify the whole S3 endpoint URL. However, I believe this is worthwhile because it enables using alternatives like minio (what I use) and S3-compatible endpoints like rackspace.

Database is configured like so:

- `TAKAHE_DATABASE_URL: "postgres://postgres:postgrese@db/takahe"`

andrewgodwin commented on takahe/settings.py:
```diff
@@ -0,0 +1,242 @@
+        case_sensitive = False
+
+
+CONFIG = Settings()
```
The internal database-driven settings system is already called Config, so this will need a different name

andrewgodwin commented on manage.py:
```diff
@@ -5,7 +5,7 @@
 # List of settings files that should guard against running certain commands
 GUARDED_ENVIRONMENTS = [
-    "production",
+    "prod",
```
I would like to keep this as "production" (and "development", etc.)

andrewgodwin commented on takahe/settings.py:
```diff
@@ -0,0 +1,242 @@
+    """
+
+    #: The default database.
+    DATABASE_URL: PostgresDsn
```
I would like the standard `PG*` environment variables to _also_ work, as they're meant to be common across software.

andrewgodwin commented on docker/docker-compose.yml:
```diff
@@ -12,11 +12,15 @@ x-takahe-common:
+      TAKAHE_EMAIL_BACKEND: "console://console"
+      TAKAHE_MAIN_DOMAIN: "example.com"
+      TAKAHE_ENVIRONMENT: "dev"
+      DJANGO_SETTINGS_MODULE: "takahe.settings"
```
I would still like the ability to have a premade development settings rather than what you've done here, as I do not use docker-compose. At minimum, that means checking an env file in that matches these.

andrewgodwin reviewed (commented):
So I actually use Pydantic settings elsewhere and I do generally like them (hell, it's used for the internal Config system in a weird way), but there's a few things here I'd like to change first.

TkTech commented:
I believe I've addressed all your comments. There's an example .venv.dev, and validation to ensure there's either a `database_url` or a `PGHOST` set. Now, if you decide to merge the sqlite support, I'd suggest changing the default database to just be sqlite instead. With that, we can tweak the defaults a little so that 0 setup is required to get started, just creating the admin user.

andrewgodwin commented:
Thanks, I'll bring this in and then make the other few tweaks I want locally rather than having you do them!
